### PR TITLE
preparing for use of `customtype` gogo extension with `NodeID` type

### PR DESCRIPTION
**Pull request by bryanchriswhite**

### CODE REVIEW

File: pkg/storj/node.go [hunk 1]
```diff
@@ -3,13 +3,174 @@
...
+			}
+
+			return uint16((i-1)*8 + zeroBits)
+		}
+	}
+
+	// NB: this should never happen
+	reason := fmt.Sprintf("difficulty matches id hash length: %d; hash (hex): % x", idLen, id)
+	zap.S().Error(reason)
+	panic(reason)
```
Comment (egonelbre) [inline pkg/storj/node.go]: This should return an error. It isn't unreasonable to get sent a NodeID that is just zeros. If it's annoying to implement right now, then just keep the panic.

File: pkg/storj/node.go [hunk 2]
```diff
@@ -3,13 +3,174 @@
...
+		return nil, err
+	}
+	return ids, nil
+}
 
-// HexString returns NodeID as hex encoded string
-func (id *NodeID) HexString() string { return hex.EncodeToString(id[:]) }
+func NodeIDFromBytes(b []byte) (NodeID, error) {
+	bLen := len(b)
+	if bLen < IdentityLength {
```
Comment (egonelbre) [inline pkg/storj/node.go]: It should match it exactly.

File: pkg/storj/node.go [hunk 3]
```diff
@@ -3,13 +3,174 @@
...
+	"math/bits"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	"golang.org/x/crypto/sha3"
+	"storj.io/storj/pkg/utils"
+)
+
+// IdentityLength is the number of bytes required to represent a node id
+const IdentityLength = 32
```
Comment (egonelbre) [inline pkg/storj/node.go]: Replace it with:
```
type NodeID [32]byte

const IdentityLength = len(NodeID{})
```

It keeps the NodeID definition clearer. However, we probably should eliminated IdentityLength altogether, it makes more sense to use `len(nodeID)`.

File: pkg/storj/node.go [hunk 4]
```diff
@@ -3,13 +3,174 @@
...
+
+var (
+	ErrNotImplemented = errs.Class("not implemented error")
+	// ErrNodeID is used when something goes wrong with a node id
+	ErrNodeID   = errs.Class("node ID error")
+	// EmptyNodeID is the zero-value for a NodeID
+	EmptyNodeID = NodeID([IdentityLength]byte{})
```
Comment (egonelbre) [inline pkg/storj/node.go]: `EmptyNodeID = NodeID{}`

File: pkg/storj/node.go [hunk 5]
```diff
@@ -3,13 +3,174 @@
...
+)
 
 // NodeID is a unique node identifier
-type NodeID [32]byte
+type NodeID [IdentityLength]byte
+type NodeIDList []NodeID
```
Comment (egonelbre) [inline pkg/storj/node.go]: Where do we need the NodeIDList definition? Also thinking whether `NodeIDs` would be nicer, but `NodeIDList` is good as well.

File: pkg/storj/node.go [hunk 6]
```diff
@@ -3,13 +3,174 @@
...
+
+func NodeIDFromString(s string) (NodeID, error) {
+	idBytes, err := base64.URLEncoding.DecodeString(s)
```
Comment (egonelbre) [inline pkg/storj/node.go]: We probably should use hex with checksum embedded or `base58check`. URLEncoding also contains `-` and `_` which make searching logs more difficult.

File: pkg/storj/node.go [hunk 7]
```diff
@@ -3,13 +3,174 @@
...
+	if bLen < IdentityLength {
+		return NodeID{}, ErrNodeID.New("not enough bytes to make a node id; have %d, need %d", bLen, IdentityLength)
+	}
+
+	var id NodeID
+	copy(id[:], b[:IdentityLength])
+	return NodeID(id), nil
+}
+
+func NodeIDFromKey(k crypto.PublicKey) (NodeID, error) {
```
Comment (egonelbre) [inline pkg/storj/node.go]: I think this part should be in provider key generation, not here.

File: pkg/storj/node.go [hunk 8]
```diff
@@ -3,13 +3,174 @@
...
+
+// String returns NodeID as hex encoded string
+func (id NodeID) String() string {
+	return base64.URLEncoding.EncodeToString(id[:])
+}
 
 // Bytes returns raw bytes of the id
 func (id NodeID) Bytes() []byte { return id[:] }
+
+func (id NodeID) Difficulty() uint16 {
```
Comment (egonelbre) [inline pkg/storj/node.go]: Is it necessary to have this as a method? The storj package will be used throughout the system, so we need to keep an eye out for things that can be elsewhere.

File: pkg/storj/node.go [hunk 9]
```diff
@@ -3,13 +3,174 @@
...
+
+func NodeIDFromString(s string) (NodeID, error) {
+	idBytes, err := base64.URLEncoding.DecodeString(s)
+	if err != nil {
+		return EmptyNodeID, ErrNodeID.Wrap(err)
+	}
+	return NodeIDFromBytes(idBytes)
+}
+
+func NodeIDsFromBytes(b [][]byte) (ids NodeIDList, err error) {
```
Comment (egonelbre) [inline pkg/storj/node.go]: Group NodeIDList things together, rather than having them mixed together with `NodeID`.

File: pkg/storj/node.go [hunk 10]
```diff
@@ -3,13 +3,174 @@
...
+
+func (n NodeIDList) Len() int {
+	return len(n)
+}
+
+func (n NodeIDList) Swap(i, j int) {
+	n[i], n[j] = n[j], n[i]
+}
+
+func (n NodeIDList) Less(i, j int) bool {
```
Comment (egonelbre) [inline pkg/storj/node.go]:
```
for k, av := range a {
	if av < b[k] {
		return true
	} else if av > b[k] {
		return false
	}
}
return false
```

File: pkg/storj/node.go [hunk 11]
```diff
@@ -3,13 +3,174 @@
...
+	return n, nil
+}
+
+func (id *NodeID) Unmarshal(data []byte) error {
+	var err error
+	*id, err = NodeIDFromBytes(data)
+	return err
+}
+
+func (id *NodeID) Size() int {
```
Comment (egonelbre) [inline pkg/storj/node.go]: Add a TODO here to remove this. We can us `len(nodeID)`

File: pkg/storj/node.go [hunk 12]
```diff
@@ -3,13 +3,174 @@
...
+	return len(id)
+}
+
+// TODO(bryanchriswhite): what should this look like?
+func (id NodeID) MarshalJSON() ([]byte, error) {
+	return nil, ErrNotImplemented.New("MarshalJSON")
```
Comment (egonelbre) [inline pkg/storj/node.go]:
```
return []byte(`"` + id.String() + `"`), nil
```

File: pkg/storj/node.go (hunk 5 above)
Comment (egonelbre) [reply pkg/storj/node.go]: Hmm, can we use `sort.Slice(nodeids, func(i, k int) bool {  return nodeids[i].Less(nodeids[k]) }` there?

File: pkg/storj/node.go (hunk 8 above)
Comment (egonelbre) [reply pkg/storj/node.go]: I mean which systems need the Difficulty? If it's just the provider, then we can move it there. But yeah, it's not too complex to keep here as well.

File: pkg/storj/node.go (hunk 10 above)
Comment (bryanchriswhite) [reply pkg/storj/node.go]: why?

File: pkg/storj/node.go (hunk 11 above)
Comment (bryanchriswhite) [reply pkg/storj/node.go]: `Size` is part of the customtype interface

File: pkg/storj/node.go (hunk 12 above)
Comment (bryanchriswhite) [reply pkg/storj/node.go]: ty

File: pkg/storj/node.go (hunk 9 above)
Comment (bryanchriswhite) [reply pkg/storj/node.go]: :+1:

File: pkg/storj/node.go (hunk 10 above)
Comment (egonelbre) [reply pkg/storj/node.go]: Faster and shorter code.

File: pkg/storj/node.go [hunk 13]
```diff
@@ -160,12 +136,17 @@ func (id *NodeID) Size() int {
...
-	return nil, ErrNotImplemented.New("MarshalJSON")
+	return []byte(`"` + id.String() + `"`), nil
 }
 
 // TODO(bryanchriswhite): what should this look like?
 func (id *NodeID) UnmarshalJSON(data []byte) error {
-	return ErrNotImplemented.New("MarshalJSON")
+	var err error
+	*id, err = NodeIDFromString(string(data))
+	if err != nil {
```
Comment (egonelbre) [inline pkg/storj/node.go]: `return err`

File: pkg/utils/lookup/lookup.go [hunk 14]
```diff
@@ -0,0 +1,26 @@
+package lookup
+
+import (
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+)
+
+// NodeIDsToLookupRequests ...
+func NodeIDsToLookupRequests(nodeIDs storj.NodeIDList) *pb.LookupRequests {
```
Comment (egonelbre) [inline pkg/utils/lookup/lookup.go]: Let's put these conversions into `pb` package. It seems to be about converting from and to pb.

File: internal/identity/identity.go [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+import (
+	"context"
+
+	"storj.io/storj/pkg/provider"
+)
+
+// helper function to generate new node identities with
+// correct difficulty and concurrency
+func NewTestIdentity() (*provider.FullIdentity, error) {
```
Comment (egonelbre) [inline internal/identity/identity.go]: There is already `provider.NewTestCA`.

File: internal/storj/node.go [hunk 16]
```diff
@@ -0,0 +1,45 @@
+package teststorj
```
Comment (egonelbre) [inline internal/storj/node.go]: package name should match the folder

File: pkg/storj/node.go (hunk 5 above)
Comment (bryanchriswhite) [reply pkg/storj/node.go]: it's used in a few places

**Comment by egonelbre** [comment]
It probably has some issues but let's deal with those separately.

**Pull request merged**